packages/doenetml-worker-rust/doenetml-core/src/component/text_input.rs

```diff
@@ -0,0 +1,516 @@
+    StateVar, StateVarInterface, StateVarMutableViewTyped, StateVarParameters,
+    StateVarReadOnlyView, StateVarReadOnlyViewTyped, StateVarTyped, StateVarValue,
+};
+use crate::utils::KeyValueIgnoreCase;
+use crate::{ComponentChild, ComponentIdx, ExtendSource};
+
+use super::{
+    ComponentEnum, ComponentNode, ComponentNodeStateVariables, ComponentProfileStateVariable,
+    RenderedComponentNode,
+};
+
+#[derive(Debug, Default, ComponentNode)]
+pub struct TextInput {
+    pub idx: ComponentIdx,
+    pub parent: Option<ComponentIdx>,
+    pub children: Vec<ComponentChild>,
+
+    pub extend: Option<ExtendSource>,
+
+    // map of descendant names to their indices
+    pub descendant_names: HashMap<String, Vec<ComponentIdx>>,
+
+    pub position: Option<DastPosition>,
+
+    pub state_variables: Vec<StateVar>,
+
+    pub rendered_state_variable_indices: Vec<usize>,
+
+    pub public_state_variable_indices: Vec<usize>,
+
+    pub state_variable_name_to_index: HashMap<String, usize>,
+
+    pub immediate_value_state_var_view: StateVarReadOnlyViewTyped<String>,
+
+    pub component_profile_state_variables: Vec<ComponentProfileStateVariable>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct TextInputRenderData {
+    pub id: usize,
+
+    pub immediate_value: String,
+}
+
+impl RenderedComponentNode for TextInput {
+    fn to_flat_dast(&mut self, _: &Vec<Rc<RefCell<ComponentEnum>>>) -> FlatDastElement {
+        let immediate_value = self
+            .immediate_value_state_var_view
+            .get_fresh_value_record_viewed()
+            .to_string();
+
+        FlatDastElement {
+            name: self.get_component_type().to_string(),
+            attributes: HashMap::new(),
+            children: vec![],
+
+            data: ElementData {
+                id: self.get_idx(),
+                action_names: Some(self.get_action_names()),
+                state: Some(HashMap::from([(
+                    "immediateValue".to_string(),
+                    StateVarValue::String(immediate_value),
+                )])),
+                ..Default::default()
+            },
+            position: self.get_position().cloned(),
+        }
+    }
+
+    fn get_flat_dast_update(&mut self) -> Option<FlatDastElementUpdate> {
+        if self
+            .immediate_value_state_var_view
+            .check_if_changed_since_last_viewed()
+        {
+            let immediate_value = self
+                .immediate_value_state_var_view
+                .get_fresh_value_record_viewed()
+                .to_string();
+
+            Some(FlatDastElementUpdate {
+                changed_attributes: None,
+                new_children: None,
+                changed_state: Some(HashMap::from([(
+                    "immediateValue".to_string(),
+                    StateVarValue::String(immediate_value),
+                )])),
+            })
+        } else {
+            None
+        }
+    }
+
+    fn get_action_names(&self) -> Vec<String> {
```

Comment:
This is used for type checking, I assume?

I'm still no sure the best way to do this, but if you set up an enum and use

```rust
use strum_macros::{EnumString, EnumVariantNames};
use strum::VariantNames;

#[derive(EnumString, EnumVariantNames)]
#[strum("camelCase")]
enum Actions {
  UpdateImmediateValue,
  UpdateValue
}
```

then you can do `Actions::VARIANTS` to get a `'static` vec of the string names.

Reply:
Deferring until another PR.